pkg/controller/framework/shared_informer.go

```diff
@@ -206,128 +167,53 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
 				if err := s.indexer.Update(d.Object); err != nil {
 					return err
 				}
-				s.processor.distribute(updateNotification{oldObj: old, newObj: d.Object})
+				s.OnUpdate(old, d.Object)
 			} else {
 				if err := s.indexer.Add(d.Object); err != nil {
 					return err
 				}
-				s.processor.distribute(addNotification{newObj: d.Object})
+				s.OnAdd(d.Object)
 			}
 		case cache.Deleted:
 			if err := s.indexer.Delete(d.Object); err != nil {
 				return err
 			}
-			s.processor.distribute(deleteNotification{oldObj: d.Object})
+			s.OnDelete(d.Object)
 		}
 	}
 	return nil
 }
 
-type sharedProcessor struct {
-	listeners []*processorListener
-}
-
-func (p *sharedProcessor) distribute(obj interface{}) {
-	for _, listener := range p.listeners {
-		listener.add(obj)
+func (p *sharedIndexInformer) OnUpdate(oldObj, newObj interface{}) {
+	for _, handler := range p.handlers {
+		handler.OnUpdate(oldObj, newObj)
```

Comment:
> This PR is not for v1.3.
> But yes, I can change it right now.

Change it how?  The current code anticipated the problem and resolved it by having a discrete queue per-listener to ensure that controllers could progress at their own rate.  There are alternatives (single rotating queue with pointers is classic), but all the ones that I'm familiar with are more difficult to implement.